third_party/python/pip.bzl

```diff
@@ -5,9 +5,121 @@
 load("@rules_python//python:pip.bzl", "pip_install")
 load("@python3//:defs.bzl", "interpreter")
 
+_WHEEL_BUILD_FILE_CONTENTS = """\
+package(default_visibility = ["//visibility:public"])
+
+exports_files([
+    "sanitized_requirements.txt",
+])
+
+filegroup(
+    name = "all_wheels",
+    srcs = glob(["**/*.whl"])
+)
+"""
+
+def _pip_wheel_impl(rctx):
+    # First, we install the Python wheel package so we can build other wheels.
+    args = [
+        rctx.path(rctx.attr.python_interpreter),
+        "-m",
+        "pip",
+        "install",
+        "wheel",
+    ]
+    progress_message = "Installing the Python wheel package"
+    rctx.report_progress(progress_message)
+    result = rctx.execute(
+        args,
+        timeout = rctx.attr.timeout,
+        quiet = rctx.attr.quiet,
+    )
+    if result.return_code:
+        fail("pip_wheel failed: {} ({})".format(result.stdout, result.stderr))
+
+    # Next, we download/build all the Python wheels for each requirement.
+    args = [
+        rctx.path(rctx.attr.python_interpreter),
+        "-m",
+        "pip",
+        "wheel",
+        "-r",
+        rctx.path(rctx.attr.requirements),
+        "-w",
+        "./",
+    ]
+    progress_message = "Pre-building Python wheels"
+    rctx.report_progress(progress_message)
+    result = rctx.execute(
+        args,
+        timeout = rctx.attr.timeout,
+        quiet = rctx.attr.quiet,
+    )
+    if result.return_code:
+        fail("pip_wheel failed: {} ({})".format(result.stdout, result.stderr))
+
+    # Last, we generate a new Python requirements file that does not contain any
+    # VCS links. We avoid just patching the existing python_requirements.txt
+    # file, and instead generate a requirements file based on the wheels that
+    # were built above. This enables us to make changes to the main
+    # python_requirements.txt file without impacting this step.
+    args = [rctx.path(rctx.attr._gen_requirements_script)]
+    progress_message = "Generating sanitzed requirements file"
+    rctx.report_progress(progress_message)
+    result = rctx.execute(
+        args,
+        timeout = rctx.attr.timeout,
+        quiet = rctx.attr.quiet,
+        working_directory = "./",
+    )
+    if result.return_code:
+        fail("pip_wheel failed: {} ({})".format(result.stdout, result.stderr))
+
+    # We need a BUILD file to load the downloaded Python packages.
+    rctx.file(
+        "BUILD.bazel",
+        _WHEEL_BUILD_FILE_CONTENTS,
+    )
+
+pip_wheel = repository_rule(
```

Comment:
I would hoist this into rules/python.bzl

Reply:
Will do in a follow up PR that deals with the detection of airgapped environment too... cuz CI hehe